nncf/ptq/data/dataloader.py

```diff
@@ -0,0 +1,118 @@
+            python object.
+        :param transform_fn: Transformation method that takes a data item
+            returned per iteration through `data_source` object and transforms
+            it to the model's expected input that can be used for the model inference.
+        :param batch_size: An integer that represents the number of consecutive elements
+            of `data_source` that were combined in a single batch.
+        """
+        self._data_source = data_source
+        self._transform_fn = transform_fn
+        self._batch_size = batch_size
+
+    @property
+    def batch_size(self) -> int:
+        """
+        Returns the number of elements return per iteration.
+
+        :return: A number of elements return per iteration.
+        """
+        return self._batch_size
+
+    def transform(self, data: DataItem) -> ModelInput:
+        return self._transform_fn(data)
+
+    def __iter__(self) -> Iterable[DataItem]:
+        return iter(self._data_source)
+
+
+def create_dataloader(data_source: DataSource,
+                      transform_fn: Callable[[DataItem], ModelInput]) -> NNCFDataLoader:
+    """
+    Wraps the provided custom data source that is an [iterable](https://docs.python.org/3/glossary.html#term-iterable)
+    python object into `NNCFDataLoader` object.
+
+    :param data_source: Custom data source that is an iterable python object.
+    :param transform_fn: The method that takes a data item returned per iteration
+        through `data_source` object and transforms it to the model's expected input
+        that can be used for the model inference.
+    :return: The object that implements the `NNCFDataLoader` interface and
+        wraps custom data source.
+    """
+    # Checks that `data_source` is an iterable python object.
+    try:
+        iter(data_source)
+    except TypeError as exc:
+        raise ValueError('The provided `data_source` object is not an iterable object. Please see '
+                         'https://docs.python.org/3/glossary.html#term-iterable to learn more '
+                         'about iterable object.') from exc
+
+    batch_size = _get_batch_size(data_source)
+
+    return NNCFDataLoaderImpl(data_source, transform_fn, batch_size)
+
+
+def _get_batch_size(data_source: DataSource) -> int:
+    """
+    Tries to determine the batch size of the provided custom data source.
+    If it was successful returns the number of elements return per iteration
+    from the `data_source` or `1` otherwise.
+
+    :param data_source: Custom data source that is an iterable python object.
+    :return: The number of elements return per iteration from the `data_source` or `1`.
+    """
+    try:
+        import torch
+    except ImportError:
+        torch = None
+
+    try:
+        import tensorflow
+    except ImportError:
+        tensorflow = None
+
+    batch_size = 1
+
+    if torch is not None and isinstance(data_source, torch.utils.data.DataLoader):
```

Comment:
I would implement a different logic when we check the interface, try to acquire the tensor, check its type and make a decision about batch size.